docs/okeps/okep-5259-no-overlay.md

```diff
@@ -0,0 +1,889 @@
+
+* For north-south traffic, we will follow the existing implementation of the BGP
+  feature; the traffic from pods will no longer be SNATed on egress. In the BGP
+  feature, SNAT is disabled for all pod egress traffic that leaves the node,
+  regardless of whether the destination route is BGP learned or not. For users
+  who only want to have no-overlay mode but do not want to expose podIPs to the
+  external, this behavior is not ideal. Instead of disabling SNAT for all the
+  pod-to-external traffic, OVN-K shall adjust its SNAT behavior so that pod IPs
+  are exposed for traffic routed via BGP, but SNAT is still applied for
+  non-BGP-routed egress traffic.
+
+* For east-west traffic, intra-node traffic (pod-to-pod, pod-to-clusterIP, and
+  pod-to-host) remains unchanged, while cross-node traffic will follow the same
+  path as north-south traffic.
+
+For the default network, we will add a new flag
+`--default-network-encapsulation` to the `ovnkube` binary. This flag accepts
+either `none` or `geneve`, with `geneve` being the default. Setting
+`--default-network-encapsulation=none` will configure the default network to
+operate in no-overlay mode. If the `--default-network-encapsulation=none` flag
+is set but the required RouterAdvertisements or FrrConfiguration CR is missing,
+`ovnkube` will exit with an error.
+
+For CUDNs, no-overlay mode shall be configured via the CUDN CRs.
+
+### API Details
+
+We introduce a new `layer3.encapsulation` field to be added to the Spec of the
+ClusterUserDefinedNetwork (CUDN) CRD. This new field will enable control over
+the network encapsulation behavior with the following options:
+
+1. **Encapsulation Configuration**:  
+   * `encapsulation`: specifies the encapsulation method (default: `Geneve`)  
+   * Supported values: `Geneve` (default) or `None` (no-overlay)
+
+The `spec` field of a ClusterUserDefinedNetwork CR is immutable. Therefore, the
+encapsulation configuration cannot be changed after the
+ClusterUserDefinedNetwork CR is created.
+
+#### Example of a layer-3 CUDN with no-overlay mode enabled
+
+A layer-3 CUDN with no-overlay mode enabled should look like this:
+
+```yaml
+apiVersion: k8s.ovn.org/v1
+kind: ClusterUserDefinedNetwork
+metadata:
+  name: my-cudn
+spec:
+  namespaceSelector:
+    matchExpressions:
+    - key: kubernetes.io/metadata.name
+      operator: In
+      values: ["red", "blue"]
+  network:
+    topology: Layer3
+    layer3:
+      role: Primary
+      mtu: 1500
+      subnets:
+      - cidr: 10.10.0.0/16
+        hostSubnet: 24
+      encapsulation: "None"
+```
+
+The topology spec for each network type is shared between CUDN and UDN. Once we
+add the new `encapsulation` field, it will be available for both CUDN and UDN.
+So, to prevent a user from setting this field in a UDN CR accidentally, a new
+`status.conditions` type `NoOverlayEnabled` will be added to CUDN/UDN CRs. If a
+user tries to create a UDN CR with no-overlay mode enabled, cluster-manager
```

Comment:
So did we dismiss the CEL in the end?

Reply:
CEL is part of the API. We want to support no-overlay UDN in the future, once BGP supports UDN. With this `status.conditions` approach, we don't need to change the API(CEL) when we support UDN. @jcaamano @tssurya WDYT?

Reply:
For me CEL still looks like the better option

Reply:
Yeah, so the CEL will be removed once we support no-overlay mode for UDN. Usually it's more confusing for a user to see their CR created and then realize that the cluster doesn't work as expected, rather than having the CR rejected from the start with a descriptive error message.

Reply:
> CEL is part of the API. We want to support no-overlay UDN in the future, once BGP supports UDN. With this `status.conditions` approach, we don't need to change the API(CEL) when we support UDN. @jcaamano @tssurya WDYT?

yea having a CEL now to be removed later is fine